src/transformStateWithClones.js

```diff
@@ -7,7 +7,38 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const states = [];
+  let currentState = { ...state };
+
+  for (const action of actions) {
+    let newState = {};
+
+    switch (action.type) {
+      case 'addProperties':
+        newState = { ...currentState, ...action.extraData };
+        break;
+
+      case 'removeProperties':
+        newState = { ...currentState };
+
+        for (const key of action.keysToRemove) {
+          delete newState[key];
+        }
+        break;
+
+      case 'clear':
+        newState = {};
+        break;
+
+      default:
+        newState = { ...currentState };
+        break;
+    }
+    states.push(newState);
+    currentState = newState;
+  }
+
+  return states;
```

Comment:
u don't need an additional object. This is redundant
u can do the same actions with ur `currentState` object
```suggestion
  const states = [];
  let currentState = { ...state };

  for (const action of actions) {
    switch (action.type) {
      case 'addProperties':
        currentState = { ...currentState, ...action.extraData };
        break;

      case 'removeProperties':
        for (const key of action.keysToRemove) {
          delete currentState[key];
        }
        break;

      case 'clear':
        newState = {};
        break;

      default:
        break;
    }
    
    states.push(newState);
  }

  return states;
```